### PR TITLE
Sync GroupLayer and Main Viewer layer orders

### LINKED ISSUE #9: [Feature] Impose `GroupLayer` order on `LayerList` to update view.
**Is your feature request related to a problem? Please describe.**

Currently reordering the items in the group layer list does not affect how the canvas appears (IE what layers appear above the others). Such changes are only reflected in updates to the LayerList interface itself.

**Describe the solution you'd like**

- [ ] A button should be added to the GroupLayersWidget that when pressed returns a "flattened" order for the Layers.
- [ ] This button should then impose this order on the existing LayerList to re-render the canvas.
- [ ] (Extension) if a group is selected, allow imposing of the relative ordering of the layers within that group on the canvas.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by willGraham01**
## Description

**What is this PR**

- [x] Bug fix
- [x] Addition of a new feature
- [ ] Other

**Why is this PR needed?**

See #9 

**What does this PR do?**

- Adds required functionality for the `GroupLayer`s widget to impose the layer order back onto the main viewer.
- Streamlines the functionality for adding and removing `Node`s and `Group`s via `add_new_{layer,group}` methods.
- Links the removal or addition of layers in the main viewer and group layers viewer, so if a LAYER item is added/removed in one, it will also be removed in the other.
- Adds tests for the functionality introduced.

It should be noted that re-ordering the layers in the main LayerList view does not change the corresponding order in the GroupLayers view. This is because:
- We envision that the GroupLayers view and standard LayerList view should not be used in tandem, and ultimate the goal is for the GroupLayers view to supersede the main viewer when integrated into the core codebase. (In which case, much of the functionality introduced in this PR will no longer be needed.) [Comment](https://github.com/brainglobe/napari-experimental/issues/9#issuecomment-2196422912)
- There is an ambiguity with regards to retaining the structure of the groups if the layers are manipulated in the main LayerList, which has no such group concepts. [Comment](https://github.com/brainglobe/napari-experimental/issues/9#issuecomment-2196403631)

## References

Closes #9 

## How has this PR been tested?

- Local testing and interaction with `napari`
- Addition of unit tests for backend functionality

## Does this PR require an update to the documentation?

Adding docstrings got sucked into #27, which targets this branch.

## Checklist:

- [ ] The code has been tested locally
- [ ] Tests have been added to cover all new functionality (unit & integration)
- [ ] The documentation has been updated to reflect any changes
- [ ] The code has been formatted with [pre-commit](https://pre-commit.com/)

### CODE REVIEW

**Review by K-Meech — APPROVED** [review]
Looks good! :tada:
Might be worth adding some quick tests to cover `_new_layer_in_main_viewer` and `on_layer_moved`?

**Comment by willGraham01** [comment]
Have added a short test for the synchronisation of layers - everything in here will likely go though if the plugin eventually replaces the main layer viewer :cry: But for the time being we can be assured that things are sync-ing each time there's a change.

File: tests/test_group_layer_widget.py [hunk 1]
```diff
@@ -66,3 +66,49 @@ def test_double_click_edit(group_layer_widget, double_click_on_view):
...
+    assert len(widget.group_layers) == 2
+
+    # Check that reordering the layer order in the group layers view
+    # forces an update in the main viewer.
+    # However, don't forget that the ordering is REVERSED in the GUI, so
+    # viewer.layers[-1] == widget.group_layers[0], etc.
+    for in_viewer, in_widget in zip(
+        reversed(viewer.layers),
+        [node.layer for node in widget.group_layers],
+        strict=True,
```
Comment (K-Meech) [inline tests/test_group_layer_widget.py]: I think `strict` was only introduced in 3.10, so you'll probably have to remove it to get the tests to pass with 3.9

**Review by K-Meech — CHANGES_REQUESTED** [review]
Looks good! Found one other `strict` to remove for the 3.9 tests to pass

**Review by K-Meech — APPROVED** [review]
Looks good :tada:

**Pull request merged**